drivers/adc/ltc2358/ltc2358.c

```diff
@@ -0,0 +1,170 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include "ltc2358.h"
+#include "errno.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Create 24-bit configuration word for the 8 channels.
+ * @param channel - Channel number.
+ * @param config_number - Configuration number for the channel.
+ * @param config_word - 24-bit config word created.
+ */
+void ltc2358_create_config_word(uint8_t channel, uint8_t config_number,
+				uint32_t *config_word)
+{
+	*config_word = *config_word | ((uint32_t)(config_number & 0x07) <<
+				       (channel * 3));
+}
+
+/**
+ * @brief Transmit 24-bit (3 bytes) of configuration information and reads back
+ * 24 bytes of data (3 bytes / 24-bit for each channel).
+ * @param dev - Device handler.
+ * @param config_word - 3 bytes of configuration data for 8 channels.
+ * @param data_array - Data array to read in 24 bytes of data from 8 channels.
+ * @return 0 in case of success, -1 otherwise.
+ */
+int32_t ltc2358_write_and_read(struct ltc2358_dev *dev, uint32_t config_word,
+			       uint8_t data_array[24])
+{
+	uint8_t tx_buff[24];
+	uint8_t i;
+
+	tx_buff[0] = (uint8_t)(config_word >> 16);
+	tx_buff[1] = (uint8_t)(config_word >> 8);
+	tx_buff[2] = (uint8_t)(config_word);
+
+	for (i = 3; i <= 23; i++)
+		tx_buff[i] = 0x00;
+
+	struct no_os_spi_msg msgs = {
+		.tx_buff = tx_buff,
+		.rx_buff = data_array,
+		.bytes_number = 24,
+		.cs_change = 1,
+	};
+
+	return no_os_spi_transfer(dev->spi_desc, &msgs, 1);
+}
+
+/**
+ * @brief Parse single channel data.
+ * @param data_array - 24 bytes of data from 8 channels.
+ * @param channel - selected channel (0-7)
+ * @return 3 bytes raw data of specific channel
+ */
+int32_t ltc2358_channel_raw_data(uint8_t data_array[24], uint8_t channel)
+{
+	union lt_union_int32_4bytes data;
+	uint8_t i;
+
+	uint32_t channel_data = 0;
+	uint8_t pos = 0;
+	data.lt_uint32 = 0;
+
+	for (i = 0; i < channel + 1; ++i) {
+		data.lt_byte[2] = data_array[pos++];
+		data.lt_byte[1] = data_array[pos++];
+		data.lt_byte[0] = data_array[pos++];
+
+		channel_data = data.lt_uint32;
+	}
+
+	return channel_data;
```

Comment:
This whole function may be replaced by `channel_data = no_os_get_unaligned_be24(&data_array[channel * 3])`. But anyway, the channel reading flow seems strange. The usual way is to define a `reg_read()` function, which is should be called in the `read_channel()`.